File: Projetos/Projeto_3ºbim/modelo/MeuTokenJWT.php
<?php
namespace Firebase\JWT;

use stdClass;
use Firebase\JWT\Key;
use Firebase\JWT\JWT;
use DomainException;
use Exception;
use InvalidArgumentException;
use UnexpectedValueException;
use Firebase\JWT\SignatureInvalidException;
use Firebase\JWT\ExpiredException;

require_once "jwt/JWT.php";
require_once "jwt/Key.php";
require_once "jwt/SignatureInvalidException.php";

require_once "jwt/ExpiredException.php";

class MeuTokenJWT
{
    //chave de criptografia, defina uma chave forte e a mantenha segura.
    private $key = "x9S4q0v+V0IjvHkG20uAxaHx1ijj+q1HWjHKv+ohxp/oK+77qyXkVj/l4QYHHTF3";

    //algoritmo de criptografia para assinatura
    //Suportados: 'HS256' , 'ES384','ES256', 'ES256K', ,'HS384', 'HS512', 'RS256', 'RS384'
    private $alg = 'HS256';
    private $type = 'JWT';
    private $iss = 'http://localhost'; //emissor do token
    private $aud = 'http://localhost'; //destinatário do token
    private $sub = "acesso_sistema";   //assunto do token
    private $iat = "";  //momento de emissão
    private $exp = "";     //momento de expiração
    private $nbf = ""; //não é válido antes do tempo especificado
    private $jti = "";  //Identificador único
    private $payload; //claims 
    //tempo de validade do token
    private $duracaoToken = 3600 * 24 * 30; //3600 segundos = 60 min

    public function gerarToken($parametro_claims)
    {

        // Criação dos headers como objeto da classe stdClass
        $objHeaders = new stdClass();
        $objHeaders->alg = $this->alg;
        $objHeaders->typ = $this->type;

        // Criação do payload como objeto da classe stdClass
        $objPayload = new stdClass();

        //Registered Claims
        $objPayload->iss = $this->iss;        // emissor do token
        $objPayload->aud = $this->aud;        // destinatário do token
        $objPayload->sub = $this->sub;        // assunto do token  
        $objPayload->iat = time();            // momento de criação do token
        $objPayload->exp = time() + $this->duracaoToken;   // momento de expiração = tempo atual + duração
        $objPayload->nbf = time();        // momento em que o token torna-se valido. 
        $objPayload->jti = bin2hex(random_bytes(16)); // gera um valor aleatório para jti;

        //Public Claims
        $objPayload->emailUsuario = $parametro_claims->emailUsuario;
        $objPayload->senhaUsuario = $parametro_claims->senhaUsuario;

        //Private claims


        // Utiliza a biblioteca do Firebase para gerar o token com os parâmetros
        $token = JWT::encode((array) $objPayload, $this->key, $this->alg, null, (array) $objHeaders);
        return $token;
    }

    public function validarToken($stringToken)
    {
        if (isset($stringToken)) {
            if ($stringToken == "") {
                return false;
            } else {
                $remover = ["Bearer ", " "];
                $token = str_replace($remover, "", $stringToken);
                try {
                    $payloadValido = JWT::decode($token, new Key($this->key, $this->alg));
                    $this->setPayload($payloadValido);
                    return true;
                } catch (SignatureInvalidException $e) {
                    // A assinatura do token é inválida.
                    //error_log("Invalid token signature: " . $e->getMessage());
                    return false;
                } catch (BeforeValidException $e) {
                    // O token não é válido ainda (antes do tempo 'nbf').
                    //error_log("Token not valid yet: " . $e->getMessage());
                    return false;
                } catch (ExpiredException $e) {
                    // O token expirou.
                    // error_log("Token expired: " . $e->getMessage());
                    return false;
                } catch (InvalidArgumentException $e) {
                    // Argumento inválido passado.
                    //error_log("Invalid argument: " . $e->getMessage());
                    return false;
                } catch (DomainException $e) {
                    // Exceção de domínio genérica.
                    //error_log("Domain exception: " . $e->getMessage());
                    return false;
                } catch (UnexpectedValueException $e) {
                    // Valor inesperado encontrado.
                    //error_log("Unexpected value: " . $e->getMessage());
                    return false;
                } catch (Exception $e) {
                    // Qualquer outra exceção genérica.
                    //error_log("General exception: " . $e->getMessage());
                    return false;
                }
            }
        }
        return false;
    }

    /**
     * Get the value of payload
     */
    public function getPayload()
    {
        return $this->payload;
    }

    /**
     * Set the value of payload
     *
     * @return  self
     */
    public function setPayload($payload)
    {
        $this->payload = $payload;

        return $this;
    }

    /**
     * Get the value of alg
     */
    public function getAlg()
    {
        return $this->alg;
    }

    /**
     * Set the value of alg
     *
     * @return  self
     */
    public function setAlg($alg)
    {
        $this->alg = $alg;

        return $this;
    }
}

File: Projetos/Projetos_4bim/flask/modelo/MeuTokenJWT.py
import jwt
import datetime
import secrets


class MeuTokenJWT:
    def __init__(self):
        self._key = "x9S4q0v+V0IjvHkG20uAxaHx1ijj+q1HWjHKv+ohxp/oK+77qyXkVj/l4QYHHTF3"  # Chave secreta
        self._alg = 'HS256'  # Algoritmo de criptografia
        self._type = 'JWT'
        self._iss = 'http://localhost'  # Emissor do token
        self._aud = 'http://localhost'  # Destinatário do token
        self._sub = "acesso_sistema"  # Assunto do token
        self._duracaoToken = 3600 * 24 * 30  # Duração do token (30 dias)
        self.payload = None


    def gerar_token(self, parametro_claims):
        headers = {
            'alg': self._alg,
            'typ': self._type
        }
        agora = datetime.datetime.now(datetime.timezone.utc)
        payload = {
            'iss': self._iss,  # Emissor do token
            'aud': self._aud,  # Destinatário do token
            'sub': self._sub,  # Assunto do token
            'iat': int(agora.timestamp()),  # Momento de criação (em segundos)
            'exp': int((agora + datetime.timedelta(seconds=self._duracaoToken)).timestamp()),  # Expiração (em segundos)
            'nbf': int(agora.timestamp()),  # Não é válido antes do tempo especificado
            'jti': secrets.token_hex(16),  # Identificador único (jti)
            'email': parametro_claims['emailUsuario'],  # Claims públicas
            'nome': parametro_claims['nomeUsuario'],  # Claims públicas

        }
        # Gera o token utilizando a biblioteca PyJWT
        token = jwt.encode(payload, self._key, algorithm=self._alg, headers=headers)
        return token


    def validar_token(self, string_token):
        if not string_token or not isinstance(string_token, str):
            print("Token inválido: o token fornecido é nulo ou não é uma string.")
            return False


        # Verifica se o token tem o prefixo 'Bearer' e o remove
        if string_token.startswith("Bearer "):
            token = string_token[len("Bearer "):].strip()
        else:
            token = string_token.strip()


        # Verifica se o token não está vazio após a limpeza
        if not token:
            print("Token inválido: o token está vazio após a remoção do prefixo.")
            return False


        try:
            # Decodifica o token
            decoded = jwt.decode(token, self._key, algorithms=[self._alg], audience=self._aud)
            self.payload = decoded
            return True
        except jwt.ExpiredSignatureError:
            print("Token expirado")
            return False
        except jwt.InvalidTokenError:
            print("Token inválido")
            return False


    def get_payload(self):
        return self.payload


    def set_payload(self, payload):
        self.payload = payload


    def get_alg(self):
        return self._alg


    def set_alg(self, alg):
        self._alg = alg
